### PR TITLE
[Mobile Payments] Charge fees on payment intent creation in Canada

### LINKED ISSUE #6708: [Mobile Payments] Charge fees on PaymentIntent creation for IPP payments in Canada
In Canada, we accept Interac payments, which are authorized and captured in a single step on the app.

For most IPP transactions, we rely on the capture step happening on the WCPay Server, so that fees (percentage charge to the merchant) can be added there. For Interac transactions, the only opportunity to add the fees is when we create the `PaymentIntent`, as there is no separate capture step.

We create the `PaymentIntent` using the Stripe Terminal SDK, i.e. from the app not from WCPay.

Since we can't know whether a customer will use Interac or another debit/credit card until after we create a `PaymentIntent`, the workaround we will use for 1.0 is to add the fees to the `PaymentIntent` when we create it, **for IPP transactions in Canada only**.

The fees are as follows:

> 2.6% + $0.25 CAD
– pdfdoF-Gj-p2#comment-1394

For non-Interac transactions, the fees we set in the app will be overridden by WCPay Server at the capture step. We've agreed to still only set these fees in Canada, to reduce any risks to the workflows of the existing merchant base in the US.

This is needed for the initial release of IPP in Canada.

See pdfdoF-Gj-p2#comment-1381 for full discussion

### PR DESCRIPTION
**Pull request by joshheald**
Closes: #6708

### Description

We charge merchants a fee for processing transactions, which is applied by WCPay Server on the capture step. 

Interac card-present payments do not do the capture step on the WCPay Server, so did not have any fees applied.

This adds the Interac fees to all payment intents in Canada. For non-Interac CA transactions, these fees are overwritten with the correct rate at the capture step as normal.

### Testing instructions

To check the fees, you'll need to open either the payment in WCPay, or the Stripe Dashboard. Here's how you find them:

**WCPay**: wp-admin > WooCommerce > Orders > {Select order} > Payment via Credit card / debit card (pi_3KltEV2EdyGr1FMV0HIiVlDl)
Payment details on WCPay

**Stripe**: You'll need access to the Stripe Dashboard for this: PdfdoF-1A-p2
Toggle `Test mode` viewing, and [find your site's account](https://dashboard.stripe.com/connect/accounts/overview) – use the filters and ⚙️ in the columns to help narrow it down.
Select `Payments` under Activity
Select the payment for your order
Payment details in Stripe Dashboard

#### Testing fees
1. Make an order or simple payment
2. Take a card-present payment for the order
3. Check the payment in the WCPay Dashboard or Stripe Dashboard – the fee should show. 

Calculate it to check it's correct! The formula is shown in WCPay, so do that calculation manually.

Repeat for the following:

- US transaction
- Canada non-Interac transaction
- Canada Interac transaction

### Screenshots

US Visa Credit Payment Fees@2x
Visa US: expected 2.6% + 0.10 = $2.70

Canada Visa Credit Payment Fees@2x
Visa CA: expected 2.6% + 0.10 = $2.70

Canada Interac Payment Fees@2x
Interac: expected 2.6% + 0.25 = $2.85

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Hardware/Hardware/CardReader/StripeCardReader/PaymentIntentParameters+Stripe.swift [hunk 1]
```diff
@@ -38,4 +42,17 @@ extension Hardware.PaymentIntentParameters {
         return returnValue
     }
 }
+
+private extension Hardware.PaymentIntentParameters {
+    enum Constants {
+        static let smallestCurrencyUnitMultiplier: Decimal = 100
```
Comment (jaclync) [inline Hardware/Hardware/CardReader/StripeCardReader/PaymentIntentParameters+Stripe.swift:48]: Just saw this part of the code the first time, and finally understood this 100 multiplier after [this Stripe doc](https://stripe.com/docs/currencies#zero-decimal). Since this could vary in other currencies, does it make sense for this multiplier to be in `CardPresentPaymentsConfiguration` (doesn't have to be this PR)?

File: WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift [hunk 2]
```diff
@@ -264,6 +269,22 @@ private extension PaymentCaptureOrchestrator {
         }
     }
 
+    private func applicationFee(for orderTotal: NSDecimalNumber, country: String) -> Decimal? {
+        guard country.uppercased() == "CA" else {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift]: nit: do you think it's worth using `SiteAddress.CountryCode.CA.rawValue` so that it's easier to look up the country with the enum's `readableCountry`?

File: WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift [hunk 3]
```diff
@@ -288,9 +309,13 @@ private extension PaymentCaptureOrchestrator {
...
 private extension PaymentCaptureOrchestrator {
     enum Constants {
-        /// Minimum order amount in USD:
+        /// Minimum order amount in USD or CAD:
         /// https://stripe.com/docs/currencies#minimum-and-maximum-charge-amounts
         static let minimumAmount = NSDecimalNumber(string: "0.5")
+
+        static let canadaFlatFee = NSDecimalNumber(string: "0.25")
+
+        static let canadaPercentageFee = NSDecimalNumber(string: "0.026")
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift:318]: Just wanted to confirm that we're using these numbers while the [WCPay doc for Canada](https://woocommerce.com/document/payments/faq/fees/#section-4) shows different values?

> 2.9% + $0.30 CAD per transaction using a domestic credit or debit card

Also, there are two other details about the fees for a credit/debit card:
> +0.6% for transactions paid using a card issued outside Canada
> [+2% for currency conversion](https://woocommerce.com/document/payments/faq/fees/currency-conversion/) when[ customer currency](https://woocommerce.com/document/payments/currencies/#section-2) is different to[ deposit currency](https://woocommerce.com/document/payments/currencies/#section-3) (USD or CAD)

Do we currently handle these two cases on the server-side, and similarly on the client-side?

File: WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift [hunk 4]
```diff
@@ -0,0 +1,225 @@
+import XCTest
+import Yosemite
+@testable import WooCommerce
+
+class PaymentCaptureOrchestratorTests: XCTestCase {
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift]: super nit:
```suggestion
final class PaymentCaptureOrchestratorTests: XCTestCase {
```

File: WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+    private var stores: MockStoresManager! = nil
+    private var sut: PaymentCaptureOrchestrator! = nil
+    private let sampleSiteID: Int64 = 1234
+
+    override func setUpWithError() throws {
+        stores = MockStoresManager(sessionManager: SessionManager.makeForTesting())
+        sut = PaymentCaptureOrchestrator(stores: stores,
+                                         paymentReceiptEmailParameterDeterminer: MockReceiptEmailParameterDeterminer())
+    }
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift]: nit: I think we usually do setup/teardown using `override func setUp() { ... }`, if it doesn't throw an error?

in any case, I think calling `super` would be nice
```suggestion
    override func setUpWithError() throws {
        super.setUpWithError()
        stores = MockStoresManager(sessionManager: SessionManager.makeForTesting())
        sut = PaymentCaptureOrchestrator(stores: stores,
                                         paymentReceiptEmailParameterDeterminer: MockReceiptEmailParameterDeterminer())
    }
```

File: WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift [hunk 6]
```diff
@@ -0,0 +1,225 @@
+import XCTest
+import Yosemite
+@testable import WooCommerce
+
+class PaymentCaptureOrchestratorTests: XCTestCase {
+
+    private var stores: MockStoresManager! = nil
+    private var sut: PaymentCaptureOrchestrator! = nil
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift]: super nit:
```suggestion
    private var stores: MockStoresManager!
    private var sut: PaymentCaptureOrchestrator!
```

File: WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    private let sampleSiteID: Int64 = 1234
+
+    override func setUpWithError() throws {
+        stores = MockStoresManager(sessionManager: SessionManager.makeForTesting())
+        sut = PaymentCaptureOrchestrator(stores: stores,
+                                         paymentReceiptEmailParameterDeterminer: MockReceiptEmailParameterDeterminer())
+    }
+
+    override func tearDownWithError() throws {
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift]: nit: similar to above, I think we usually call `override func tearDown() { ... }`. Also, we might want to reset `stores` and `sut` to `nil` here too, followed by a call to?

File: WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        }
+    }
+
+}
+
+extension PaymentCaptureOrchestratorTests {
+    func mockCollectPaymentActionReaderMessage() {
+        stores.whenReceivingAction(ofType: CardPresentPaymentAction.self) { action in
+            if case let .collectPayment(_, _, _, onCardReaderMessage, _, _) = action {
+                onCardReaderMessage(.waitingForInput("Present card"))
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift]: just curious, why is this necessary?

File: WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+        switch action {
+        case .collectPayment(siteID: _, orderID: _, parameters: let parameters, onCardReaderMessage: _, onProcessingCompletion: _, onCompletion: _):
+            assertEqual(expectedFee, parameters.applicationFee)
+        default:
+            XCTFail("Collecting Payment did not send collectPayment CardPresentPaymentAction")
+        }
+    }
+
+    func test_collectPayment_for_a_payment_to_a_CA_gateway_account_rounds_down_applicationFees_to_2dp_where_next_digit_is_below_5() throws {
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift:155]: nice tests! 👏

**Review by jaclync — APPROVED** [review]
Thanks for handling this challenging case and p2'ing the issue! Code LGTM 👍 

One non-blocking question from testing:
- CA store / WCPay: Interac ✅ (2.6% + $0.25)
- CA store / WCPay: non-Interac ✅ (2.6% + $0.10)
- US store / Stripe: the rate in the [Stripe dashboard](https://user-images.githubusercontent.com/2472348/165479306-539db3fb-df03-4e42-b323-6e536e8e2051.png) doesn't seem to match 2.6%+$0.1. I couldn't find the doc on the fees, do we take any fees from the Stripe extension❓ 
- US store / WCPay: ✅  (2.6% + $0.10)

File: Hardware/Hardware/CardReader/StripeCardReader/PaymentIntentParameters+Stripe.swift (hunk 1 above)
Comment (joshheald) [reply Hardware/Hardware/CardReader/StripeCardReader/PaymentIntentParameters+Stripe.swift:48]: Yep that would definitely be a good place for it! I've added #6739

File: WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift (hunk 3 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift:318]: >Just wanted to confirm that we're using these numbers while the [WCPay doc for Canada](https://woocommerce.com/document/payments/faq/fees/#section-4) shows different values?

The numbers I used were the numbers provided by @kalessil that WCPay use, but we're getting them confirmed with Stripe. I've left them matching WCPay's implementation until we hear differently.

> Do we currently handle these two cases on the server-side, and similarly on the client-side?

We don't handle them at all on Interac payments, but Interac cards are all Canadian-issued, so these scenarios don't apply. 

For all other payments they would/should be handled on the server side... but they're likely to be very rare anyway because the apps don't really support multi-currency. I think it might be possible to reproduce it with an order placed on the web for a multi-currency site and then paid for as CoD.

File: WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift (hunk 8 above)
Comment (joshheald) [reply WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift]: It means that we wait until we've built the `paymentParameters`, which contain the `applicationFee` in an accessible way (we can't access it on the PaymentIntent directly), before doing our asserts.

[The action is sent here](https://github.com/woocommerce/woocommerce-ios/blob/1e9cd52dbdc926b1a5cbbf2af9c024e9065591b8/WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift#L79)

And then we grab the action from the stores in each test, to get the parameters out of it.

However... there's a better way to do it which explaining all this made me notice! New commit incoming.

File: WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift]: Done in 9607f86e
I want to extract that to a standalone in Yosemite now too, like we did for `CurrencyCode`, as it's not really just about a `SiteAddress` and we're using it in a context which has nothing to do with `SiteAddress`, but it's still probably better than the plain string.

**Comment by joshheald** [comment]
Thanks for the review @jaclync!

> * US store / Stripe: the rate in the [Stripe dashboard](https://user-images.githubusercontent.com/2472348/165479306-539db3fb-df03-4e42-b323-6e536e8e2051.png) doesn't seem to match 2.6%+$0.1. I couldn't find the doc on the fees, do we take any fees from the Stripe extension❓

No, the Stripe extension uses the merchant's own Stripe account, so we **can't** take fees AFAIK. They're an option for "Connect" accounts in Stripe, which we use in WCPay as sub-accounts of the overall WCPay account, one per merchant.

The image you linked looks like my WCPay CA test site, so has fees charged, but you wouldn't see that on a Stripe extension site.

File: WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift (hunk 8 above)
Comment (joshheald) [reply WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/PaymentCaptureOrchestratorTests.swift]: Improved tests in e7fe1ad6

**Comment by joshheald** [comment]
> No, the Stripe extension uses the merchant's own Stripe account, so we can't take fees AFAIK. They're an option for "Connect" accounts in Stripe, which we use in WCPay as sub-accounts of the overall WCPay account, one per merchant.

@jaclync This is something we should check if the workaround is still in place when we do Stripe Extension in Canada... what happens with these fees for the non-WCPay accounts, do they break anything, etc. I've added a task to the follow on project:
- #6741
